src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -908,33 +921,61 @@ protected long logEvent(String eventType, JSONObject eventProperties, JSONObject
-            event.put("device_manufacturer", replaceWithJSONNull(deviceInfo.getManufacturer()));
-            event.put("device_model", replaceWithJSONNull(deviceInfo.getModel()));
-            event.put("carrier", replaceWithJSONNull(deviceInfo.getCarrier()));
-            event.put("country", replaceWithJSONNull(deviceInfo.getCountry()));
-            event.put("language", replaceWithJSONNull(deviceInfo.getLanguage()));
-            event.put("platform", platform);
             event.put("uuid", UUID.randomUUID().toString());
             event.put("sequence_number", getNextSequenceNumber());
 
+            if (trackingOptions.shouldTrackVersionName()) {
+                event.put("version_name", replaceWithJSONNull(deviceInfo.getVersionName()));
+            }
+            if (trackingOptions.shouldTrackOsName()) {
+                event.put("os_name", replaceWithJSONNull(deviceInfo.getOsName()));
+            }
+            if (trackingOptions.shouldTrackOsVersion()) {
+                event.put("os_version", replaceWithJSONNull(deviceInfo.getOsVersion()));
+            }
+            if (trackingOptions.shouldTrackDeviceBrand()) {
+                event.put("device_brand", replaceWithJSONNull(deviceInfo.getBrand()));
+            }
+            if (trackingOptions.shouldTrackDeviceManufacturer()) {
+                event.put("device_manufacturer", replaceWithJSONNull(deviceInfo.getManufacturer()));
+            }
+            if (trackingOptions.shouldTrackDeviceModel()) {
+                event.put("device_model", replaceWithJSONNull(deviceInfo.getModel()));
+            }
+            if (trackingOptions.shouldTrackCarrier()) {
+                event.put("carrier", replaceWithJSONNull(deviceInfo.getCarrier()));
+            }
+            if (trackingOptions.shouldTrackCountry()) {
+                event.put("country", replaceWithJSONNull(deviceInfo.getCountry()));
+            }
+            if (trackingOptions.shouldTrackLanguage()) {
+                event.put("language", replaceWithJSONNull(deviceInfo.getLanguage()));
+            }
+            if (trackingOptions.shouldTrackPlatform()) {
+                event.put("platform", platform);
+            }
+
             JSONObject library = new JSONObject();
             library.put("name", Constants.LIBRARY);
             library.put("version", Constants.VERSION);
             event.put("library", library);
 
             apiProperties = (apiProperties == null) ? new JSONObject() : apiProperties;
-            Location location = deviceInfo.getMostRecentLocation();
-            if (location != null) {
-                JSONObject locationJSON = new JSONObject();
-                locationJSON.put("lat", location.getLatitude());
-                locationJSON.put("lng", location.getLongitude());
-                apiProperties.put("location", locationJSON);
+            JSONObject apiPropertiesTrackingOptionsCopy = Utils.cloneJSONObject(apiPropertiesTrackingOptions);
```

Comment:
api properties was already cloned before calling `logEvent`, so should be safe to mutate.
We don't want to mutate the `apiPropertiesTrackingOptions` blob in memory since that gets copied to each event.